libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerProvider.kt

```diff
@@ -1,28 +1,70 @@
 package com.mapbox.navigation.core.routerefresh
 
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
 import com.mapbox.navigation.base.route.RouteRefreshOptions
 import com.mapbox.navigation.core.PrimaryRouteProgressDataProvider
 import com.mapbox.navigation.core.RoutesProgressDataProvider
 import com.mapbox.navigation.core.directions.session.DirectionsSession
 import com.mapbox.navigation.core.ev.EVDynamicDataHolder
 import com.mapbox.navigation.core.ev.EVRefreshDataProvider
 import com.mapbox.navigation.core.routealternatives.AlternativeMetadataProvider
+import com.mapbox.navigation.utils.internal.ThreadController
+import com.mapbox.navigation.utils.internal.Time
 import java.util.Date
 
 internal object RouteRefreshControllerProvider {
 
+    @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
     fun createRouteRefreshController(
+        threadController: ThreadController,
         routeRefreshOptions: RouteRefreshOptions,
         directionsSession: DirectionsSession,
         primaryRouteProgressDataProvider: PrimaryRouteProgressDataProvider,
         alternativeMetadataProvider: AlternativeMetadataProvider,
         evDynamicDataHolder: EVDynamicDataHolder,
-    ) = RouteRefreshController(
-        routeRefreshOptions,
-        directionsSession,
-        RoutesProgressDataProvider(primaryRouteProgressDataProvider, alternativeMetadataProvider),
-        EVRefreshDataProvider(evDynamicDataHolder),
-        DirectionsRouteDiffProvider(),
-        { Date() },
-    )
+    ): RouteRefreshControllerImpl {
+        val scope = threadController.getMainScopeAndRootJob().scope
+        val routeRefresher = RouteRefresher(
+            RoutesProgressDataProvider(
+                primaryRouteProgressDataProvider,
+                alternativeMetadataProvider
+            ),
+            EVRefreshDataProvider(evDynamicDataHolder),
+            DirectionsRouteDiffProvider(),
+            directionsSession,
+        )
+        val routeRefresherExecutor = RouteRefresherExecutor(
+            routeRefresher,
+            scope,
+            routeRefreshOptions.intervalMillis
+        )
+        val stateHolder = RouteRefreshStateHolder()
+        val refreshObserversManager = RefreshObserversManager()
+        val routeRefresherResultProcessor = RouteRefresherResultProcessor(
+            refreshObserversManager,
+            ExpiringDataRemover { Date() },
+            Time.SystemImpl,
+            routeRefreshOptions.intervalMillis * (PlannedRouteRefreshController.MAX_RETRY_COUNT + 1)
```

Comment:
```suggestion
            routeRefreshOptions.intervalMillis * 3
```
The process for clearing out-of-date information is now based purely on time, it's only artificially coupled with number of attempts because there could be many more attempts happening with immediate refresh being taken into account.

Could we remove the concept of attempts from the codebase in this case? This would in some way simplify this already complex setup and the only side effect is that a `start/cancel/success/failure` `RouteRefreshStatesObserver` callbacks would be invoke for each request instead of being invoked every third request.

Reply:
> and the only side effect is that a start/cancel/success/failure RouteRefreshStatesObserver callbacks would be invoked for each request instead of being invoked every third request.

I was trying to avoid breaking this behaviour. IMO this is a breaking change.

Reply:
The retry logic is an entirely internal construct, we don't specify that behavior anywhere in the docs or the public API. The only entry in public, stable `RouteRefreshOptions` docs is:
```
In case of a route refresh failing for `3 * `[intervalMillis] (default is 15 min), expired incidents and congestion annotations will be removed from the route...
```

On top of that, the current `REFRESH_STATE_FINISHED_FAILED` doc says:
```
The state becomes [REFRESH_STATE_FINISHED_FAILED] when a route refresh failed and the route is cleaned up of expired data, see [RouteRefreshOptions] for details.
```
which is not true for this PR because a `REFRESH_STATE_FINISHED_FAILED` will also be delivered for a failed immediate request which might not have cleaned up expired data yet.

I think we should consider refactoring the `REFRESH_STATE_FINISHED_FAILED` doc (since it's experimental) to delete the expired data removal language, and expose a new `REFRESH_STATE_CLEARED_EXPIRED` to be deliver after a failure that also exceeded the time threshold (or some other alternative to this).

Reply:
Actually, if I just change `PlannedRouteRefreshController.MAX_RETRY_COUNT + 1` to 3, route refresh state callbacks behaviour won't change.

Reply:
> I think we should consider refactoring the REFRESH_STATE_FINISHED_FAILED doc (since it's experimental) to delete the expired data removal language, and expose a new REFRESH_STATE_CLEARED_EXPIRED to be deliver after a failure that also exceeded the time threshold (or some other alternative to this).

I like that!